Sourcery/Sourcery.swift

```diff
@@ -347,36 +348,59 @@ extension Sourcery {
 // MARK: - Generation
 extension Sourcery {
 
-    fileprivate func generate(_ templatePaths: Paths, output: Path, parsingResult: ParsingResult) throws {
+    fileprivate func generate(source: Source, templatePaths: Paths, output: Output, parsingResult: ParsingResult) throws {
         Log.info("Loading templates...")
         let allTemplates = try templates(from: templatePaths)
         Log.info("Loaded \(allTemplates.count) templates.")
 
         Log.info("Generating code...")
         status = ""
 
-        guard output.isDirectory else {
+        if output.path.isDirectory {
+            try allTemplates.forEach { template in
+                let result = try generate(template, forParsingResult: parsingResult, outputPath: output.path)
+                let outputPath = output.path + generatedPath(for: template.sourcePath)
+                try self.output(result: result, to: outputPath)
+
+                if let linkTo = output.linkTo {
+                    try link(outputPath, to: linkTo)
+                }
+            }
+        } else {
             let result = try allTemplates.reduce("") { result, template in
-                return result + "\n" + (try generate(template, forParsingResult: parsingResult))
+                return result + "\n" + (try generate(template, forParsingResult: parsingResult, outputPath: output.path))
             }
+            try self.output(result: result, to: output.path)
 
-            try self.output(result: result, to: output)
-            return
+            if let linkTo = output.linkTo {
+                try link(output.path, to: linkTo)
+            }
         }
 
-        try allTemplates.forEach { template in
-            let outputPath = output + generatedPath(for: template.sourcePath)
-            let result = try generate(template, forParsingResult: parsingResult)
-
-            try self.output(result: result, to: outputPath)
+        if let linkTo = output.linkTo {
+            try linkTo.project.writePBXProj(path: linkTo.projectPath)
         }
 
         Log.info("Finished.")
     }
 
+    private func link(_ output: Path, to linkTo: Output.LinkTo) throws {
+        if let target = linkTo.project.target(named: linkTo.target) {
+            let fileGroup = linkTo.project.addGroup(named: linkTo.group ?? "SourceryGenerated", to: linkTo.project.rootGroup)
+            try linkTo.project.addSourceFile(at: output, toGroup: fileGroup, target: target, sourceRoot: linkTo.projectPath)
+        }
+    }
+
     private func output(result: String, to outputPath: Path) throws {
+        var result = result
         if !result.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
-            try writeIfChanged(Sourcery.generationHeader + result, to: outputPath)
+            if outputPath.extension == "swift" {
+                result = Sourcery.generationHeader + result
+            }
+            if !outputPath.parent().exists {
```

Comment:
whats going to happen if you auto link a file in --watch mode and then the template is nil and the file gets pruned? or do you have a guard for that

Reply:
Hm, good catch. It will keep the file linked to the project so it will not build any more. So I guess we should "unlink" it when it is being deleted.

Reply:
@krzysztofzablocki it looks cumbersome to implement though as we don't store file reference anywhere and then will have to go through all the project files to find one to remove. I will just add the check for file existence so that it does not add files that are removed.

Reply:
Actually it is [already there](https://github.com/xcodeswift/xcproj/blob/master/Sources/xcproj/PBXProjObjects%2BHelpers.swift#L102) in xcodeproj methods so, I guess no need to do anything on our side